### PR TITLE
[X86] Remove knl/knm specific ISAs supports

### PR DESCRIPTION
**Pull request by FreddyLeaf**
Cont. patch after https://github.com/llvm/llvm-project/pull/75580

### CODE REVIEW

**Comment by FreddyLeaf** [comment]
gcc recently formally removed knl/knm supports at trunk: https://godbolt.org/z/nKvo1P48d
Per conclusion before in https://github.com/llvm/llvm-project/pull/75580, this patch removed specific ISA supports of knl/knm while keep -march/tune support

**Comment by phoebewang** [comment]
Please note it in release notes.

File: llvm/lib/Target/X86/X86InstrAVX512.td [hunk 1]
```diff
@@ -9265,6 +9265,33 @@ multiclass avx512_fp28_s<bits<8> opc, string OpcodeStr,X86VectorVTInfo _,
...
+}
+
+multiclass avx512_eri_s_ass<bits<8> opc, string OpcodeStr> {
+  defm SSZ : avx512_fp28_s_ass<opc, OpcodeStr#"ss", f32x_info>,
+             EVEX_CD8<32, CD8VT1>, VEX_LIG, T8, PD, EVEX, VVVV;
+  defm SDZ : avx512_fp28_s_ass<opc, OpcodeStr#"sd", f64x_info>,
+             EVEX_CD8<64, CD8VT1>, VEX_LIG, REX_W, T8, PD, EVEX, VVVV;
+}
+
+defm VRCP28   : avx512_eri_s_ass<0xCB, "vrcp28">;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAVX512.td:9292]: Why still keep these instructions?

File: llvm/lib/Target/X86/X86InstrAVX512.td [hunk 2]
```diff
@@ -9325,6 +9345,43 @@ multiclass avx512_fp28_p_sae<bits<8> opc, string OpcodeStr, X86VectorVTInfo _,
...
+multiclass  avx512_eri_ass<bits<8> opc, string OpcodeStr> {
+   defm PSZ : avx512_fp28_p_ass<opc, OpcodeStr#"ps", v16f32_info>,
+              avx512_fp28_p_sae_ass<opc, OpcodeStr#"ps", v16f32_info>,
+              T8, PD, EVEX_V512, EVEX_CD8<32, CD8VF>;
+   defm PDZ : avx512_fp28_p_ass<opc, OpcodeStr#"pd", v8f64_info>,
+              avx512_fp28_p_sae_ass<opc, OpcodeStr#"pd", v8f64_info>,
+              T8, PD, EVEX_V512, REX_W, EVEX_CD8<64, CD8VF>;
+}
+
+defm VRSQRT28 : avx512_eri_ass<0xCC, "vrsqrt28">, EVEX;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAVX512.td:9381]: ditto.

File: llvm/test/tools/llvm-mca/X86/Generic/resources-avx512er.s [hunk 3]
```diff
@@ -1,373 +0,0 @@
-# NOTE: Assertions have been autogenerated by utils/update_mca_test_checks.py
```
Comment (phoebewang) [inline llvm/test/tools/llvm-mca/X86/Generic/resources-avx512er.s:1]: Why removing this?

File: llvm/test/CodeGen/X86/unfoldMemoryOperand.mir [hunk 4]
```diff
@@ -23,7 +23,7 @@
...
+  attributes #0 = { nofree norecurse nosync nounwind uwtable writeonly mustprogress "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+x87,-aes,-avx,-avx2,-avx512bf16,-avx512bitalg,-avx512bw,-avx512cd,-avx512dq,-avx512f,-avx512ifma,-avx512pf,-avx512vbmi,-avx512vbmi2,-avx512vl,-avx512vnni,-avx512vp2intersect,-avx512vpopcntdq,-avxvnni,-f16c,-fma,-fma4,-gfni,-kl,-pclmul,-sha,-sse,-sse2,-sse3,-sse4.1,-sse4.2,-sse4a,-ssse3,-vaes,-vpclmulqdq,-widekl,-xop" "tune-cpu"="generic" }
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/unfoldMemoryOperand.mir:26]: I think maybe we can do an NFC patch to remove these "target-features"="..." given we have already had "target-cpu"="xxx".

File: llvm/test/CodeGen/X86/avx512-gather-scatter-intrin-deprecated.ll [hunk 5]
```diff
@@ -268,30 +268,6 @@ define void @gather_qps(<8 x i64> %ind, <8 x float> %src, ptr %base, ptr %stbuf)
   ret void
 }
 
-declare  void @llvm.x86.avx512.gatherpf.qps.512(i8, <8 x i64>, ptr , i32, i32);
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/avx512-gather-scatter-intrin-deprecated.ll:271]: Surprise to see they were working without a avx512pf feature.

File: llvm/test/CodeGen/X86/insert-prefetch-invalid-instr.ll [hunk 6]
```diff
@@ -8,16 +8,12 @@ target triple = "x86_64-unknown-linux-gnu"
...
   ret i32 291, !dbg !11
 }
 
 ; Function Attrs: inaccessiblemem_or_argmemonly nounwind
 declare void @llvm.prefetch(ptr nocapture readonly, i32, i32, i32) #1
 
-; Function Attrs: argmemonly nounwind
-declare void @llvm.x86.avx512.gatherpf.dpd.512(i8, <8 x i32>, ptr, i32, i32) #2
-
 attributes #0 = {"target-cpu"="x86-64" "target-features"="+avx512pf,+sse4.2,+ssse3"}
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/insert-prefetch-invalid-instr.ll:17]: Remove `+avx512pf` or the whole `"target-features"="..."`

File: llvm/test/CodeGen/X86/unfoldMemoryOperand.mir (hunk 4 above)
Comment (FreddyLeaf) [reply llvm/test/CodeGen/X86/unfoldMemoryOperand.mir:26]: Here target-cpu is `x86-64`, but the target-features are not strictly equal to `x86-64`'s features

File: llvm/test/CodeGen/X86/avx512-gather-scatter-intrin-deprecated.ll (hunk 5 above)
Comment (FreddyLeaf) [reply llvm/test/CodeGen/X86/avx512-gather-scatter-intrin-deprecated.ll:271]: yes, this should be a bug before.

File: llvm/test/tools/llvm-mca/X86/Generic/resources-avx512er.s (hunk 3 above)
Comment (FreddyLeaf) [reply llvm/test/tools/llvm-mca/X86/Generic/resources-avx512er.s:1]: scheduleinfo was removed. Now I think we should keep schedule info for these instructions, will add back.

File: llvm/test/CodeGen/X86/insert-prefetch-invalid-instr.ll (hunk 6 above)
Comment (FreddyLeaf) [reply llvm/test/CodeGen/X86/insert-prefetch-invalid-instr.ll]: [13a0f8d](https://github.com/llvm/llvm-project/pull/92883/commits/13a0f8d87343fd1a6380a66b7986ad1ce34b06fc)

**Comment by FreddyLeaf** [comment]
> Please note it in release notes.

[13a0f8d](https://github.com/llvm/llvm-project/pull/92883/commits/13a0f8d87343fd1a6380a66b7986ad1ce34b06fc)

**Comment by phoebewang** [comment]
LGTM but I'd like @RKSimon to take a second look.

File: llvm/docs/ReleaseNotes.rst [hunk 7]
```diff
@@ -139,6 +139,9 @@ Changes to the Windows Target
 Changes to the X86 Backend
 --------------------------
 
+- Removed knl/knm specific ISA lowerings: AVX512PF, AVX512ER, PREFETCHWT1,
```
Comment (RKSimon) [inline llvm/docs/ReleaseNotes.rst:142]: "ISA intrinsics" instead?

**Review by RKSimon — APPROVED** [review]
LGTM with one minor

File: llvm/docs/ReleaseNotes.rst (hunk 7 above)
Comment (FreddyLeaf) [reply llvm/docs/ReleaseNotes.rst]: [3470095](https://github.com/llvm/llvm-project/pull/92883/commits/347009534afb057b9c33f66cc76888f302e62c8c)

**Pull request merged**